examples/throughput/subscriber.cpp

```diff
@@ -0,0 +1,323 @@
+ * whether to send data in bursts. The publisher will continue to send data forever
+ * unless a time out is specified. The subscriber will receive data and output the
+ * total amount received and the data rate in bytes per second. It will also indicate
+ * if any samples were received out of order. A maximum number of cycles can be
+ * specified and once this has been reached the subscriber will terminate and output
+ * totals and averages.
+ */
+
+#define BYTES_PER_SEC_TO_MEGABITS_PER_SEC 125000
+#define MAX_SAMPLES 1000
+#define subprefix "=== [Subscriber] "
+
+using namespace org::eclipse::cyclonedds;
+
+static unsigned long long outOfOrder(0); /*keeps track of out of order samples*/
+
+static unsigned long long total_bytes(0); /*keeps track of total bytes received*/
+
+static unsigned long long total_samples(0); /*keeps track of total samples received*/
+
+static std::chrono::milliseconds pollingDelay(-1); /*default is a listener*/
+
+static unsigned long long maxCycles(0); /*maximum number of display cycles to show*/
+
+static std::string partitionName("Throughput example"); /*name of the domain on which the throughput test is run*/
+
+static std::map<dds::core::InstanceHandle, unsigned long long> mp; /*collection of expected sequence numbers*/
+
+static volatile sig_atomic_t done(false); /*semaphore for keeping track of whether to run the test*/
+
+static int parse_args(int argc, char **argv)
+{
+  /*
+   * Get the program parameters
+   * Parameters: subscriber [maxCycles] [pollingDelay] [partitionName]
+   */
+  if (argc == 2 && (strcmp (argv[1], "-h") == 0 || strcmp (argv[1], "--help") == 0))
+  {
+    std::cout << subprefix << "Usage (parameters must be supplied in order):\n" <<
+                 subprefix << "./subscriber [maxCycles (0 = infinite)] [pollingDelay (ms, 0 = waitset, -1 = listener)] [partitionName]\n" <<
+                 subprefix << "Defaults:\n" <<
+                 subprefix << "./subscriber 0 0 \"Throughput example\"\n" << std::flush;
+    return EXIT_FAILURE;
+  }
+
+  if (argc > 1)
+  {
+    maxCycles = static_cast<unsigned long long>(atoi (argv[1])); /* The number of times to output statistics before terminating */
+  }
+  if (argc > 2)
+  {
+    pollingDelay = std::chrono::milliseconds(atoi(argv[2])); /* The number of ms to wait between reads (0 = waitset, -1 = listener) */
+  }
+  if (argc > 3)
+  {
+    partitionName = argv[3]; /* The name of the partition */
+  }
+  return EXIT_SUCCESS;
+}
+
+unsigned long long do_take(dds::sub::DataReader<ThroughputModule::DataType>& rd)
+{
+  auto samples = rd.take();
+  unsigned long long valid_samples = 0;
+  for (const auto & s:samples)
+  {
+    if (!s.info().valid())
+      continue;
+
+    auto ih = s.info().publication_handle();
+    auto ct = s.data().count();
+    auto it = mp.find(ih);
+    /*check whether the received sequence number matches that which we expect*/
+    if (it != mp.end() &&
+        it->second != ct)
+      outOfOrder++;
```

Comment:
Is the intention to show out of order samples as a rolling total? Or should it get reset between batches of samples?

Reply:
It was implemented this way because the C throughput example also shows cumulative received samples, and I thought it would be better to have the C++ throughput example to mirror this approach and head off possible questions about why C and C++ give different output :)

Reply:
I'd agree with keeping it the same as the C example (and with that also with the OSPL-alternatives as those examples are also used 'in combinations' to compare both DDS implementations)

Reply:
I'm all for keeping things consistent. But the expected behaviour would be per batch (especially when it is part of a function that operates per batch), and not a rolling total that needs to be broken apart to see the numbers per batch.

If it was a tally of the out of order *batches* then it makes sense, but this is a count of the out of order *samples* in a single batch. If you miss any samples, it will look like you miss the same number of samples for every batch following that batch.

Reply:
ok